### PR TITLE
Remove ruby-filemagic dependency and refactor/redesign content type white/blacklisting feature

### PR DESCRIPTION
**Pull request by mehlah**
This PR removes `CarrierWave::MagicMimeTypes` processor module at d6fa112 (see #1813 for background).

As discussed with @thomasfedb (https://github.com/carrierwaveuploader/carrierwave/pull/1813#issuecomment-168316910), it also remove the dependency on `ruby-filemagic` in white/blacklist content types (b5bec3d).

Background: Before this, CarrierWave was relying on mime-types or ruby-filemagic
gems to get/set the content type of a file in an inconsistent way.
After making mime-types a runtime dependency for CW,
`CarrierWave::MimeTypes` has been dropped, and `CarrierWave::MagicMimeTypes` as well with this.

White/blacklisting content types feature (introduced by #1596) was relying only on ruby-filemagic to read the content type of a file.
It was optional and required mixing a module to the uploader.

This PR removes this dependency by using the already available content type (thanks to `Mime::Types`).
This let us drop any dependency on ruby-filemagic in CarrierWave.
It makes JRuby support better as ruby-filemagic is unsupported on JRuby due to the C extension.

I also made some refactors, renamed methods like `whitelist_mime_type_pattern` to `content_type_whitelist`, cleaned up specs and docs.
The module doesn't need to be optional now, so it's included by default (like `ExtensionWhitelist` and `ExtensionBlacklist`)

To keep it consistent with the current way of white/blacklisting extensions, `content_type_whitelist` now accepts an array of strings or regexps to white/blacklist content types. I can move 07c80cb to a separate PR if you prefer.

### CODE REVIEW

File: spec/uploader/callback_spec.rb [hunk 1]
```diff
@@ -3,19 +3,27 @@
 describe CarrierWave::Uploader do
 
   it "should keep callbacks on different classes isolated" do
+    default_before_callbacks = [
+      :check_whitelist!,
```
Comment (thomasfedb) [inline spec/uploader/callback_spec.rb]: Might be worth changing `check_whitelist!` to `check_extension_whitelist!` and similar. Happy to do it as a separate PR though.

Comment (mehlah) [inline spec/uploader/callback_spec.rb]: I actually did it already :bowtie:
https://github.com/mehlah/carrierwave/compare/remove-magicmime...mehlah:refactor-extensions-white/blacklist#diff-2bca1d8b1d06f3d29a6553468c7b940fR7
I'll PR the refactor for better consistency as soon as this is accepted.

Comment (thomasfedb) [inline spec/uploader/callback_spec.rb]: :+1:

**Comment by mehlah** [comment]
I screwed up 6df8c8341be7c87870dedd328e74b88f7a7268a9 after a rebase (made the build red https://travis-ci.org/carrierwaveuploader/carrierwave/jobs/99844432). Now fixed ;)
